linera-chain/src/block_tracker.rs

```diff
@@ -0,0 +1,220 @@
+        view: &mut SystemExecutionStateView<C>,
+        context: ChainExecutionContext,
+    ) -> Result<(), ChainError>
+    where
+        C: Context + Clone + Send + Sync + 'static,
+    {
+        self.next_message_index = txn_outcome.next_message_index;
+        self.next_application_index = txn_outcome.next_application_index;
+        self.next_chain_index = txn_outcome.next_chain_index;
+        self.oracle_responses
+            .push(txn_outcome.oracle_responses.clone());
+        self.events.push(txn_outcome.events.clone());
+        self.blobs.push(txn_outcome.blobs.clone());
+        self.messages.push(txn_outcome.outgoing_messages.clone());
+        self.operation_results
+            .push(OperationResult(txn_outcome.operation_result.clone()));
+
+        let mut resource_controller = self.resource_controller.with_state(view).await?;
+
+        for message_out in &txn_outcome.outgoing_messages {
+            resource_controller
+                .track_message(&message_out.message)
+                .with_execution_context(context)?;
+        }
+
+        resource_controller
+            .track_block_size_of(&(
+                &txn_outcome.oracle_responses,
+                &txn_outcome.outgoing_messages,
+                &txn_outcome.events,
+                &txn_outcome.blobs,
+            ))
+            .with_execution_context(context)?;
+
+        for blob in &txn_outcome.blobs {
+            if blob.content().blob_type() == BlobType::Data {
+                resource_controller
+                    .track_blob_published(blob.content())
+                    .with_execution_context(context)?;
+            }
+        }
+
+        self.resource_controller
+            .track_block_size_of(&(&txn_outcome.operation_result))
+            .with_execution_context(context)?;
+
+        self.transaction_index += 1;
+        Ok(())
+    }
+
+    /// Returns recipient chain ids for outgoing messages in the block.
+    pub fn recipients(&self) -> BTreeSet<ChainId> {
+        self.messages
+            .iter()
+            .flatten()
+            .map(|msg| msg.destination)
+            .collect()
+    }
+
+    /// Returns the execution context for the current transaction.
+    pub fn chain_execution_context(&self, transaction: &Transaction<'_>) -> ChainExecutionContext {
+        match transaction {
+            Transaction::ReceiveMessages(_) => {
+                ChainExecutionContext::IncomingBundle(self.transaction_index)
+            }
+            Transaction::ExecuteOperation(_) => {
+                ChainExecutionContext::Operation(self.transaction_index)
+            }
+        }
+    }
+
+    /// Returns a mutable reference to the resource controller.
+    pub fn resource_controller_mut(
+        &mut self,
+    ) -> &mut ResourceController<Option<AccountOwner>, ResourceTracker> {
+        self.resource_controller
+    }
+
+    /// Finalized the execution and returns the collected results.
```

Comment:
```suggestion
    /// Finalizes the execution and returns the collected results.
```

Reply:
Done in [Fix typos](https://github.com/linera-io/linera-protocol/pull/3982/commits/9e15aa7a4b38afcebcb24a8a1ecead5edf8ad988)